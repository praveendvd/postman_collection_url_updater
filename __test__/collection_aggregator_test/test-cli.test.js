const exec = require('shelljs').exec,
  SDK = require('postman-collection'),
  FS = require('fs');

describe('Validate main.js test', () => {
  describe('Validate cli output messages', () => {
    it('Should show help if arguments are missing', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js"', { silent: true });
      expect(commandResponse.stderr).toMatchSnapshot();
    });

    it('Should show help if -c argument is missing', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -r "test" -w "{{baseURL}}/{{path}}" -s "new_collection.json"', { silent: true });
      expect(commandResponse.stderr).toMatchSnapshot();
      expect(commandResponse.stdout).toMatchSnapshot();
    });

    it('Should show help if -r argument is missing', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -c "test/collection/collection.json" -w "{{baseURL}}/{{path}}" -s "new_collection.json"', { silent: true });
      expect(commandResponse.stderr).toMatchSnapshot();
      expect(commandResponse.stdout).toMatchSnapshot();
    });

    it('Should show help if -w argument is missing', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -c "test/collection/collection.json" -r "{{baseURL}}/{{path}}" -s "new_collection.json"', { silent: true });
      expect(commandResponse.stderr).toMatchSnapshot();
      expect(commandResponse.stdout).toMatchSnapshot();
    });

    it('Should show file not found error if collection doesnt exists', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -c "test/collection/collection.json" -r "https://localhost:23456/api/v1/{{path}}" -w "{{baseURL}}/{{path}}" -s "new_collection.json"', { silent: true });
      expect(commandResponse.stderr).toMatchSnapshot();
      expect(commandResponse.stdout).toMatchSnapshot();
    });

    it('Should show file saved message correctly with default path', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -c "__test__/url_replacer_test/collection/collection.json" -r "https://localhost:23456/api/v1/{{path}}" -w "{{baseURL}}/{{path}}"', { silent: true });
      expect(commandResponse.stderr).toMatch('');
      expect(commandResponse.stdout).toMatch(/File saved to: .*[\\|\/]__test__[\\|\/]url_replacer_test[\\|\/]collection[\\|\/]new_collection.json/g);
    });

    it('Should show file saved message correctly with custom path', async () => {
      const commandResponse = exec('node "bin/mainCollectionAggregator.js" -c "__test__/url_replacer_test/collection/collection.json" -r "https://localhost:23456/api/v1/{{path}}" -w "{{baseURL}}/{{path}}" -s "__test__/url_replacer_test/collection/output/new.json"', { silent: true });
      expect(commandResponse.stderr).toMatch('');
      expect(commandResponse.stdout).toMatch(/File saved to: .*[\\|\/]__test__[\\|\/]url_replacer_test[\\|\/]collection[\\|\/]output[\\|\/]new\.json/g);
    });
  })

  describe('Validate collection changes', () => {
    let collectionPath = '__test__/url_replacer_test/collection/collection.json',
      outputCollectionPath = '__test__/url_replacer_test/collection/output/new.json';

    it('validate no changes happens when there is no match', async () => {

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "https://localhost:23456/api/v1/{{path}}" -w "{{baseURL}}/{{path}}" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();
      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });

    it('validate change happens only for urls other properties remains the same', async () => {

      function updatedRequestObject(source, depthID, postionID) {
        source.request.url.host = ["{{baseURL}}"];
        source.request.url.path = [
          "{{path}}",
          "testpath",
          `:pathvariable${depthID}-${postionID}`,
          "path",
          `request${depthID}-${postionID}`,
        ];
        delete source.request.url.protocol
      }

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "https://www.testdomain0.ie" -w "{{baseURL}}/{{path}}" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();

      updatedRequestObject(sourceCollection.item[0].item[0], 1, 1);
      updatedRequestObject(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0], 'n', 1);
      updatedRequestObject(sourceCollection.item[2], 0, 1);

      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });

    it('validate change protocol is shown if not removed', async () => {

      function updatedRequestObject(source, depthID, postionID) {
        source.request.url.host = ["{{baseURL}}"];
        source.request.url.path = [
          "{{path}}",
          "testpath",
          `:pathvariable${depthID}-${postionID}`,
          "path",
          `request${depthID}-${postionID}`,
        ];
      }

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "www.testdomain0.ie" -w "{{baseURL}}/{{path}}" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();

      updatedRequestObject(sourceCollection.item[0].item[0], 1, 1);
      updatedRequestObject(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0], 'n', 1);
      updatedRequestObject(sourceCollection.item[2], 0, 1);

      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });


    it('validate path varaible is removed if replaced', async () => {

      function updatedRequestObject(source, depthID, postionID) {
        source.request.url.host = ["{{baseURL}}"];
        source.request.url.path = [
          "{{path}}",
          "path",
          `request${depthID}-${postionID}`,
        ];
        source.request.url.variable = []
      }

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "www.testdomain0.ie/testpath/:pathvariable1-1" -w "{{baseURL}}/{{path}}" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();

      updatedRequestObject(sourceCollection.item[0].item[0], 1, 1);

      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });


    it('validate query parameters are updated', async () => {

      function updatedRequestObject(source, depthID, postionID) {
        source.request.url.path = [
          "testpath",
          `:pathvariable${depthID}-${postionID}`,
          "path",
          `request${depthID}-${postionID}`,
        ];
        source.request.url.query = [
          { key: 'query1-1-0', value: 'queryvalue1-1-0-0' },
          { key: 'test', value: 'queryvalue1-1-0' },
          { key: 'query1-1-1', value: 'queryvalue1-1-1' }
        ]

      }

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "/:pathvariable1-1/path/request1-1?query1-1-0" -w "/:pathvariable1-1/path/request1-1?query1-1-0=queryvalue1-1-0-0&test" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();

      updatedRequestObject(sourceCollection.item[0].item[0], 1, 1);

      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });


    it('validate query and path variables are removed if replaced completely', async () => {

      function updatedRequestObject(source, depthID, postionID) {
        source.request.url.path = [
          'v1',
          '{{path}}'
        ];
        source.request.url.variable = [];
        source.request.url.query = [];
        source.request.url.host = ["{{baseURL}}"];
        source.request.url.protocol = "{{protocol}}";

      }

      const sourceCollection = new SDK.Collection(JSON.parse(FS.readFileSync(collectionPath).toString())).toJSON();
      const commandResponse = exec(`node "bin/mainCollectionAggregator.js" -c "${collectionPath}" -r "https://www.testdomain0.ie/testpath/:pathvariable1-1/path/request1-1?query1-1-0=queryvalue1-1-0&query1-1-1=queryvalue1-1-1" -w "{{protocol}}://{{baseURL}}/v1/{{path}}" -s "${outputCollectionPath}"`, { silent: true });
      const outputCollection = new SDK.Collection(JSON.parse(FS.readFileSync(outputCollectionPath).toString())).toJSON();

      updatedRequestObject(sourceCollection.item[0].item[0], 1, 1);

      expect(outputCollection.item[0].item[0].request).toStrictEqual(sourceCollection.item[0].item[0].request);
      expect(outputCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request)
        .toStrictEqual(sourceCollection.item[1].item[0].item[0].item[0].item[0].item[0].item[0].item[0].request);
      expect(outputCollection.item[2].request).toStrictEqual(sourceCollection.item[2].request);
      expect(outputCollection.item[3].request).toStrictEqual(sourceCollection.item[3].request);
    });

  })
})


