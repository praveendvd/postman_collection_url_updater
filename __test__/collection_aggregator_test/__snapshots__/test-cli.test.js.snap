// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate main.js test Validate cli output messages Should show error if collection file is non existing 1`] = `
"ENOENT: no such file or directory, open '__test__\\\\collection_aggregator_test\\\\collection\\\\collection5.json'
"
`;

exports[`Validate main.js test Validate cli output messages Should show error if collection file is non existing 2`] = `""`;

exports[`Validate main.js test Validate cli output messages Should show error if collection file is not valid 1`] = `
"Invalid/corrupted collection provided. Please provide path to valid source collection
"
`;

exports[`Validate main.js test Validate cli output messages Should show error if collection file is not valid 2`] = `""`;

exports[`Validate main.js test Validate cli output messages Should show error if collection folder is not folder 1`] = `
"ENOTDIR: not a directory, scandir '__test__\\\\collection_aggregator_test\\\\__snapshots__\\\\test-cli-unit.test.js.snap'
"
`;

exports[`Validate main.js test Validate cli output messages Should show error if collection folder is not folder 2`] = `""`;

exports[`Validate main.js test Validate cli output messages Should show error if collection folder is not found 1`] = `
"ENOENT: no such file or directory, scandir '__test__\\\\collection_aggregator_test\\\\nonexisting'
"
`;

exports[`Validate main.js test Validate cli output messages Should show error if collection folder is not found 2`] = `""`;

exports[`Validate main.js test Validate cli output messages Should show help if arguments are missing 1`] = `
"
Usage: -c \\"root/collection_name.json\\" -r \\"{{Baseurl}}/path1/path2\\" -w
\\"{{Baseurl}}/{{path}}\\" -s \\"root/new_collection_name.json\\"

or

Usage: -i

Options:
      --help                  Show help                                [boolean]
      --version               Show version number                      [boolean]
  -i, --interactive           Pass -i to enter Interactive mode
                                                      [boolean] [default: false]
  -n, --new_collection_name   Name of the new collection
                                     [string] [default: \\"Aggregated_collection\\"]
  -l, --collection_list       Pass relative or absolute collection file paths
                                                                         [array]
  -d, --collection_directory  Directory containing all collection files [string]
  -s, --save_as               path to save new collection
      [string] [default: \\"currentDirectory/<newcollectionName>.collection.json\\"]

Examples:
  mainCollectionAggregator.js -n            - Using collection list
  \\"new_collection\\" -l
  \\"path/collection1.json\\"
  \\"path/collection1.json\\"
  \\"path/collection1.json\\"
  mainCollectionAggregator.js -d            - Using collection directory
  \\"./collection_directory\\"

Please provide either collection_directory or collection_list
[Missing required property -l or -d]
"
`;

exports[`Validate main.js test Validate cli output messages Should show help if arguments are missing 2`] = `""`;
